src/pymatching/sparse_blossom/driver/user_graph.test.cc

```
@@ -166,3 +181,144 @@ TEST(UserGraph, DecodeUserGraphDetectionEventOnBoundaryNode) {
+    EXPECT_EQ(handler.handled_errors[0], (HandledError{0.125, 1, 2, {0}}));
+}
+
+// Test an error that flips multiple logical observables.
+TEST(IterDemInstructionsTest, ErrorWithMultipleObservables) {
+    stim::DetectorErrorModel dem("error(0.3) D3 D4 L1 L3");
+    TestHandler handler;
+    pm::iter_dem_instructions_include_correlations(dem, handler);
+    ASSERT_EQ(handler.handled_errors.size(), 1);
+    EXPECT_EQ(handler.handled_errors[0], (HandledError{0.3, 3, 4, {1, 3}}));
+}
+
+// Test an error with probability 0. It should be ignored.
+TEST(IterDemInstructionsTest, ZeroProbabilityError) {
+    stim::DetectorErrorModel dem("error(0.0) D0 D1");
+    TestHandler handler;
+    pm::iter_dem_instructions_include_correlations(dem, handler);
+    ASSERT_TRUE(handler.handled_errors.empty());
+}
+
+// Test an error involving more than two detectors (a hyperedge). This should be ignored.
+TEST(IterDemInstructionsTest, ThreeDetectorErrorIsIgnored) {
+    stim::DetectorErrorModel dem("error(0.1) D0 D1 D2");
+    TestHandler handler;
+    pm::iter_dem_instructions_include_correlations(dem, handler);
+    ASSERT_TRUE(handler.handled_errors.empty());
+}
+
+// Test a decomposed error instruction. The handler should be called for each component.
+TEST(IterDemInstructionsTest, DecomposedError) {
+    stim::DetectorErrorModel dem("error(0.1) D0 D1 ^ D2 D3 L0 ^ D4");
+    TestHandler handler;
+    pm::iter_dem_instructions_include_correlations(dem, handler);
+    ASSERT_EQ(handler.handled_errors.size(), 3);
+
+    EXPECT_EQ(handler.handled_errors[0], (HandledError{0.1, 0, 1, {}}));
+    EXPECT_EQ(handler.handled_errors[1], (HandledError{0.1, 2, 3, {0}}));
+    EXPECT_EQ(handler.handled_errors[2], (HandledError{0.1, 4, SIZE_MAX, {}}));
+}
+
+// Test a decomposed error where one of the components is a hyperedge and should be ignored.
+TEST(IterDemInstructionsTest, DecomposedErrorWithIgnoredComponent) {
+    stim::DetectorErrorModel dem("error(0.15) D0 D1 ^ D2 D3 D4 ^ D5 D6 L2");
```

As mentioned above, this is fine but might be slightly better to throw an exception if an error is decomposed (contains at least one `^`) but has a component that is a hyperedge.

---

The suggestion makes sense; will do in a follow up since I have a few PRs coming in that are based off of this branch and it will be cleaner